# Packages

In [1]:
import pandas as pd

# Importation des données

In [2]:
df = pd.read_csv("2025-08-QD.csv")

In [3]:
df.head()

sasdate     GDPC1   PCECC96    PCDGx     PCESVx     PCNDx   GPDIC1  \
0    factors     0.000     0.000   1.0000     1.0000    1.0000    0.000   
1  transform     5.000     5.000   5.0000     5.0000    5.0000    5.000   
2   3/1/1959  3352.129  2039.017  68.6651  1374.1739  689.1172  354.894   
3   6/1/1959  3427.667  2070.508  71.2495  1394.7096  694.8197  382.500   
4   9/1/1959  3430.057  2092.138  72.6270  1413.6627  697.4699  357.798   

       FPIx  Y033RC1Q027SBEAx     PNFIx  ...  TNWMVBSNNCBBDIx   TLBSNNBx  \
0    0.0000            1.0000    1.0000  ...             0.00       0.00   
1    5.0000            5.0000    5.0000  ...             2.00       5.00   
2  357.0174           47.8021  171.0783  ...          1341.70  333245.24   
3  368.0640           49.2332  176.0004  ...          1294.17  345845.01   
4  371.9585           50.8372  180.7406  ...          1320.75  354808.05   

   TLBSNNBBDIx    TABSNNBx  TNWBSNNBx  TNWBSNNBBDIx     CNCFx  S&P 500  \
0         0.00        0.00       0.00          0.00    0.0000   1.0000   
1         1.00        5.00       5.00          2.00    5.0000   5.0000   
2    266668.11  2426352.41    2092.54       1674.48  124.9663  55.5167   
3    273329.36  2434975.21    2088.55       1650.63  126.5305  57.5067   
4    275139.48  2434604.03    2079.21       1612.34  128.9557  58.7300   

   S&P div yield  S&P PE ratio  
0         0.0000        0.0000  
1         2.0000        5.0000  
2         3.1765       18.6211  
3         3.1012       19.2899  
4         3.0720       18.9544  

[5 rows x 246 columns]

We have 127 variables. We need to verify if the date est de 1959 to 2025 par ordre de mensualité. After that, let's check if our interested variables are on the dataset. 

# 2. Verify the frequence and the date of time series

In [4]:
# Drop the row of transform
df = df.iloc[2:].reset_index(drop=True)
print(df.head())

     sasdate     GDPC1   PCECC96    PCDGx     PCESVx     PCNDx   GPDIC1  \
0   3/1/1959  3352.129  2039.017  68.6651  1374.1739  689.1172  354.894   
1   6/1/1959  3427.667  2070.508  71.2495  1394.7096  694.8197  382.500   
2   9/1/1959  3430.057  2092.138  72.6270  1413.6627  697.4699  357.798   
3  12/1/1959  3439.832  2094.495  69.1573  1430.8157  701.7617  369.442   
4   3/1/1960  3517.181  2114.532  71.3815  1443.7405  703.5338  406.581   

       FPIx  Y033RC1Q027SBEAx     PNFIx  ...  TNWMVBSNNCBBDIx   TLBSNNBx  \
0  357.0174           47.8021  171.0783  ...          1341.70  333245.24   
1  368.0640           49.2332  176.0004  ...          1294.17  345845.01   
2  371.9585           50.8372  180.7406  ...          1320.75  354808.05   
3  367.6452           50.7342  180.2333  ...          1165.67  349400.81   
4  379.9162           52.4985  186.4967  ...          1327.65  362821.95   

   TLBSNNBBDIx    TABSNNBx  TNWBSNNBx  TNWBSNNBBDIx     CNCFx  S&P 500  \
0    266668.11  24

In [7]:
# Let us make the date as index
df.sasdate
df.sasdate = pd.to_datetime(df.sasdate)
df = df.set_index(df.sasdate)
df = df.drop("sasdate", axis=1)
df.head()

GDPC1   PCECC96    PCDGx     PCESVx     PCNDx   GPDIC1  \
sasdate                                                                 
1959-03-01  3352.129  2039.017  68.6651  1374.1739  689.1172  354.894   
1959-06-01  3427.667  2070.508  71.2495  1394.7096  694.8197  382.500   
1959-09-01  3430.057  2092.138  72.6270  1413.6627  697.4699  357.798   
1959-12-01  3439.832  2094.495  69.1573  1430.8157  701.7617  369.442   
1960-03-01  3517.181  2114.532  71.3815  1443.7405  703.5338  406.581   

                FPIx  Y033RC1Q027SBEAx     PNFIx     PRFIx  ...  \
sasdate                                                     ...   
1959-03-01  357.0174           47.8021  171.0783  282.9707  ...   
1959-06-01  368.0640           49.2332  176.0004  292.8479  ...   
1959-09-01  371.9585           50.8372  180.7406  287.3778  ...   
1959-12-01  367.6452           50.7342  180.2333  279.2502  ...   
1960-03-01  379.9162           52.4985  186.4967  287.8417  ...   

            TNWMVBSNNCBBDIx   TLBSNNBx  TLBSNNBBDIx    TABSNNBx  TNWBSNNBx  \
sasdate                                                                      
1959-03-01          1341.70  333245.24    266668.11  2426352.41    2092.54   
1959-06-01          1294.17  345845.01    273329.36  2434975.21    2088.55   
1959-09-01          1320.75  354808.05    275139.48  2434604.03    2079.21   
1959-12-01          1165.67  349400.81    254967.79  2429055.21    2079.05   
1960-03-01          1327.65  362821.95    278692.55  2451394.44    2087.97   

            TNWBSNNBBDIx     CNCFx  S&P 500  S&P div yield  S&P PE ratio  
sasdate                                                                   
1959-03-01       1674.48  124.9663  55.5167         3.1765       18.6211  
1959-06-01       1650.63  126.5305  57.5067         3.1012       19.2899  
1959-09-01       1612.34  128.9557  58.7300         3.0720       18.9544  
1959-12-01       1517.14  137.0372  57.7633         3.1572       17.8221  
1960-03-01       1603.82  130.1872  56.2767         3.3850       16.8284  

[5 rows x 245 columns]

In [8]:
# Beginning and the end of the time series
print(df.index[0])
print(df.index[-1])

1959-03-01 00:00:00
2025-06-01 00:00:00


We observe that the dataset is vary 1959 to 2025. We normaly find the frequence of quaterly.

In [9]:
pd.infer_freq(df.index)

'QS-DEC'

The quarterly periods starting in December. 

# 3. Analyze missing values

In [12]:
df.isna().sum().sum()

np.int64(3027)

We have 3027 missing values.